fog/overseer/server/src/worker.rs

```diff
@@ -0,0 +1,544 @@
+                        logger,
+                    )
+                })
+                .expect("Could not spawn OverseerWorkerThread"),
+        );
+
+        Self {
+            join_handle,
+            stop_requested,
+        }
+    }
+
+    /// Stop and join the db poll thread
+    pub fn stop(&mut self) -> Result<(), ()> {
+        if let Some(join_handle) = self.join_handle.take() {
+            self.stop_requested.store(true, Ordering::SeqCst);
+            join_handle.join().map_err(|_| ())?;
+        }
+
+        Ok(())
+    }
+}
+
+impl Drop for OverseerWorker {
+    fn drop(&mut self) {
+        let _ = self.stop();
+    }
+}
+
+/// The thread that performs the Fog Overseer logic.
+struct OverseerWorkerThread<DB: RecoveryDb> {
+    /// The list of FogIngestClients that Overseer uses to communicate with
+    /// each node in the Fog Ingest cluster that it's monitoring.
+    ingest_clients: Vec<FogIngestGrpcClient>,
+
+    /// The database that contains, among other things, info on the Fog Ingest
+    /// cluster's ingress keys.
+    recovery_db: DB,
+
+    /// If this is true, the worker will not perform it's monitoring logic.
+    is_enabled: Arc<AtomicBool>,
+
+    /// If this is true, the thread will stop.
+    stop_requested: Arc<AtomicBool>,
+
+    logger: Logger,
+}
+
+/// This associates an IngestSummary with an IngestClient. This makes it easy
+/// to query a given node based on its IngestSummary.
+struct IngestSummaryNodeMapping {
+    node_index: usize,
+
+    ingest_summary: IngestSummary,
+}
+
+impl<DB: RecoveryDb> OverseerWorkerThread<DB>
+where
+    OverseerError: From<DB::Error>,
+{
+    /// Poll the Fog Ingest cluster every 5 seconds.
+    const POLLING_FREQUENCY: Duration = Duration::from_secs(5);
+
+    /// Try a request to Fog Ingest node this many times if you encounter an
+    /// error.
+    const NUMBER_OF_TRIES: u64 = 3;
+
+    pub fn start(
+        ingest_clients: Vec<FogIngestGrpcClient>,
+        recovery_db: DB,
+        is_enabled: Arc<AtomicBool>,
+        stop_requested: Arc<AtomicBool>,
+        logger: Logger,
+    ) {
+        let thread = Self {
+            ingest_clients,
+            recovery_db,
+            is_enabled,
+            stop_requested,
+            logger,
+        };
+        thread.run();
+    }
+
+    fn run(self) {
+        loop {
+            log::trace!(self.logger, "Overseer worker start of thread.");
+            std::thread::sleep(Self::POLLING_FREQUENCY);
+
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Overseer worker thread stopping.");
+                break;
+            }
+
+            if !self.is_enabled.load(Ordering::SeqCst) {
+                log::trace!(self.logger, "Overseer worker is currently disabled.");
+                continue;
+            }
+
+            let ingest_summary_node_mappings: Vec<IngestSummaryNodeMapping> =
+                self.retrieve_ingest_summary_node_mappings();
```

Comment:
I would want to make it so that if some nodes are unreachable, we back off until the next polling cycle -- so this would check if any node gave an error, then we log it, and `continue` the loop

Reply:
Done.